Fail to export genesis from zero-height

#### Specifications

Juno version: `v16.0.0-4-gb34af55`
OS & Version: OSX
Commit hash: b34af555ba5918f1aec0aa8c58df7ed89bb653aa

#### Expected behavior
The command export --for-zero-height is used to export the genesis file from the genesis time when the chain launched.

#### Actual behavior
 It failed due to an invalid validator address retrieval panic error.

#### Steps to reproduce the behavior
1. run the node via `make test-node`
2. stop the node after the node runs up normally to reach a height greater than zero
3. export the genesis file with command `junod export --for-zero-height`
4. the process exit with error panic:"panic: expected validator, not found" 

#### Backtrace

````
$junod export --for-zero-height --home "/Users/mingwang/.juno1"
4:05PM INF asserting crisis invariants inv=1/12 module=x/crisis name=bank/nonnegative-outstanding
4:05PM INF asserting crisis invariants inv=2/12 module=x/crisis name=bank/total-supply
4:05PM INF asserting crisis invariants inv=3/12 module=x/crisis name=transfer/total-escrow-per-denom
4:05PM INF asserting crisis invariants inv=4/12 module=x/crisis name=gov/module-account
4:05PM INF asserting crisis invariants inv=5/12 module=x/crisis name=distribution/nonnegative-outstanding
4:05PM INF asserting crisis invariants inv=6/12 module=x/crisis name=distribution/can-withdraw
4:05PM INF asserting crisis invariants inv=7/12 module=x/crisis name=distribution/reference-count
4:05PM INF asserting crisis invariants inv=8/12 module=x/crisis name=distribution/module-account
4:05PM INF asserting crisis invariants inv=9/12 module=x/crisis name=staking/module-accounts
4:05PM INF asserting crisis invariants inv=10/12 module=x/crisis name=staking/nonnegative-power
4:05PM INF asserting crisis invariants inv=11/12 module=x/crisis name=staking/positive-delegation
4:05PM INF asserting crisis invariants inv=12/12 module=x/crisis name=staking/delegator-shares
4:05PM INF asserted all invariants duration=3.389292 height=4 module=x/crisis
panic: expected validator, not found

goroutine 1 [running]:
github.com/CosmosContracts/juno/v17/app.(*App).prepForZeroHeightGenesis(_, {{0x105846670, 0x140000580b8}, {0x10585b480, 0x14002468140}, {{0x0, 0x0}, {0x0, 0x0}, 0x4, ...}, ...}, ...)
        github.com/CosmosContracts/juno/v17/app/export.go:158 +0xf38
github.com/CosmosContracts/juno/v17/app.(*App).ExportAppStateAndValidators(0x14000ffb000, 0x1, {0x106d9a7a8, 0x0, 0x0}, {0x106d9a7a8, 0x0, 0x0})
        github.com/CosmosContracts/juno/v17/app/export.go:27 +0x118
github.com/CosmosContracts/juno/v17/cmd/junod/cmd.appCreator.appExport({{{0x105851ab0, 0x140001e0d20}, {0x1058665a0, 0x14000970050}, {0x105857f00, 0x14000aa3bc0}, 0x140004a8558}}, {0x105847be0, 0x1400117f930}, {0x10585b990, ...}, ...)
        github.com/CosmosContracts/juno/v17/cmd/junod/cmd/root.go:350 +0x144
github.com/cosmos/cosmos-sdk/server.ExportCmd.func1(0x14000166c00, {0x14000ea4180?, 0x0?, 0x3?})
        github.com/cosmos/cosmos-sdk@v0.47.4/server/export.go:73 +0x338
github.com/spf13/cobra.(*Command).execute(0x14000166c00, {0x14000ea40f0, 0x3, 0x3})
        github.com/spf13/cobra@v1.7.0/command.go:940 +0x5c4
github.com/spf13/cobra.(*Command).ExecuteC(0x1400124a300)
        github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.7.0/command.go:992
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.7.0/command.go:985
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x140000021a0?, {0x104627212, 0x5}, {0x14000054618, 0x15})
        github.com/cosmos/cosmos-sdk@v0.47.4/server/cmd/execute.go:32 +0x13c
main.main()
        github.com/CosmosContracts/juno/v17/cmd/junod/main.go:18 +0x44
````
````

When submitting logs: please submit them as text and not screenshots.

Hey @Reecepbcups, the issue existed in the release of v16.0.0 as well. 
This issue was first introduced from `ignite CLI` or `starport` when scaffolding the chain. Where the root cause lies in the incorrect validator address fetch via the `addr := sdk.ValAddress(iter.Key()[1:])`, which will lead to a panic error when this command is executed.
Note that the validator address is fetched via `iter.Key()[1:]`, which excludes the `ValidatorKey`, but it still includes the prefix, the length of the address. This incomplete path leads to the error: `expected validator, not found`   

Please see the log as below:
```
$ junod version 
v16.0.0

# mingwang @ Mings-MacBook-Pro in ~/codeme/src/github.com/CosmosContracts/juno on git:v16.0.0 x [23:04:39] 
$ junod export --for-zero-height --home "/Users/mingwang/.juno1" 
11:05PM INF asserting crisis invariants inv=1/12 module=x/crisis name=distribution/nonnegative-outstanding
11:05PM INF asserting crisis invariants inv=2/12 module=x/crisis name=distribution/can-withdraw
11:05PM INF asserting crisis invariants inv=3/12 module=x/crisis name=distribution/reference-count
11:05PM INF asserting crisis invariants inv=4/12 module=x/crisis name=distribution/module-account
11:05PM INF asserting crisis invariants inv=5/12 module=x/crisis name=gov/module-account
11:05PM INF asserting crisis invariants inv=6/12 module=x/crisis name=staking/module-accounts
11:05PM INF asserting crisis invariants inv=7/12 module=x/crisis name=staking/nonnegative-power
11:05PM INF asserting crisis invariants inv=8/12 module=x/crisis name=staking/positive-delegation
11:05PM INF asserting crisis invariants inv=9/12 module=x/crisis name=staking/delegator-shares
11:05PM INF asserting crisis invariants inv=10/12 module=x/crisis name=transfer/total-escrow-per-denom
11:05PM INF asserting crisis invariants inv=11/12 module=x/crisis name=bank/nonnegative-outstanding
11:05PM INF asserting crisis invariants inv=12/12 module=x/crisis name=bank/total-supply
11:05PM INF asserted all invariants duration=5.454292 height=5 module=x/crisis
panic: expected validator, not found

goroutine 1 [running]:
github.com/CosmosContracts/juno/v16/app.(*App).prepForZeroHeightGenesis(_, {{0x1054fa210, 0x1400019c008}, {0x10550ee40, 0x1400237b8c0}, {{0x0, 0x0}, {0x0, 0x0}, 0x5, ...}, ...}, ...)
        github.com/CosmosContracts/juno/v16/app/export.go:158 +0xf38
github.com/CosmosContracts/juno/v16/app.(*App).ExportAppStateAndValidators(0x140008fd000, 0x1, {0x106a41c78, 0x0, 0x0}, {0x106a41c78, 0x0, 0x0})
        github.com/CosmosContracts/juno/v16/app/export.go:27 +0x118
github.com/CosmosContracts/juno/v16/cmd/junod/cmd.appCreator.appExport({{{0x105505530, 0x140000acf80}, {0x105519f60, 0x14001033730}, {0x10550b920, 0x140006e3640}, 0x14000198ba8}}, {0x1054fb7b8, 0x1400106f320}, {0x10550f350, ...}, ...)
        github.com/CosmosContracts/juno/v16/cmd/junod/cmd/root.go:350 +0x144
github.com/cosmos/cosmos-sdk/server.ExportCmd.func1(0x1400130b500, {0x14000eb1e90?, 0x0?, 0x3?})
        github.com/cosmos/cosmos-sdk@v0.47.3/server/export.go:73 +0x338
github.com/spf13/cobra.(*Command).execute(0x1400130b500, {0x14000eb1e00, 0x3, 0x3})
        github.com/spf13/cobra@v1.7.0/command.go:940 +0x5c4
github.com/spf13/cobra.(*Command).ExecuteC(0x140012cd500)
        github.com/spf13/cobra@v1.7.0/command.go:1068 +0x340
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.7.0/command.go:992
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.7.0/command.go:985
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x140000021a0?, {0x1042e0fc6, 0x5}, {0x140001a87e0, 0x15})
        github.com/cosmos/cosmos-sdk@v0.47.3/server/cmd/execute.go:32 +0x13c
main.main()
        github.com/CosmosContracts/juno/v16/cmd/junod/main.go:18 +0x44

```

--- a/app/export.go
+++ b/app/export.go
@@ -2,6 +2,7 @@ package app
 
 import (
 	"encoding/json"
+	"fmt"
 	"log"
 
 	tmproto "github.com/cometbft/cometbft/proto/tendermint/types"
@@ -71,20 +72,21 @@ func (app *App) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddrs []str
 
 	// withdraw all validator commission
 	app.AppKeepers.StakingKeeper.IterateValidators(ctx, func(_ int64, val stakingtypes.ValidatorI) (stop bool) {
-		_, err := app.AppKeepers.DistrKeeper.WithdrawValidatorCommission(ctx, val.GetOperator())
-		if err != nil {
-			panic(err)
-		}
+		_, _ = app.AppKeepers.DistrKeeper.WithdrawValidatorCommission(ctx, val.GetOperator())
 		return false
 	})
 
 	// withdraw all delegator rewards
 	dels := app.AppKeepers.StakingKeeper.GetAllDelegations(ctx)
 	for _, delegation := range dels {
-		_, err := app.AppKeepers.DistrKeeper.WithdrawDelegationRewards(ctx, delegation.GetDelegatorAddr(), delegation.GetValidatorAddr())
+		valAddr, err := sdk.ValAddressFromBech32(delegation.ValidatorAddress)
 		if err != nil {
 			panic(err)
 		}
+
+		delAddr := sdk.MustAccAddressFromBech32(delegation.DelegatorAddress)
+
+		_, _ = app.AppKeepers.DistrKeeper.WithdrawDelegationRewards(ctx, delAddr, valAddr)
 	}
 
 	// clear validator slash events
@@ -105,20 +107,28 @@ func (app *App) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddrs []str
 		feePool.CommunityPool = feePool.CommunityPool.Add(scraps...)
 		app.AppKeepers.DistrKeeper.SetFeePool(ctx, feePool)
 
-		err := app.AppKeepers.DistrKeeper.Hooks().AfterValidatorCreated(ctx, val.GetOperator())
-
-		return err != nil
+		if err := app.AppKeepers.DistrKeeper.Hooks().AfterValidatorCreated(ctx, val.GetOperator()); err != nil {
+			panic(err)
+		}
+		return false
 	})
 
 	// reinitialize all delegations
 	for _, del := range dels {
-		err := app.AppKeepers.DistrKeeper.Hooks().BeforeDelegationCreated(ctx, del.GetDelegatorAddr(), del.GetValidatorAddr())
+		valAddr, err := sdk.ValAddressFromBech32(del.ValidatorAddress)
 		if err != nil {
 			panic(err)
 		}
-		err = app.AppKeepers.DistrKeeper.Hooks().AfterDelegationModified(ctx, del.GetDelegatorAddr(), del.GetValidatorAddr())
-		if err != nil {
-			panic(err)
+		delAddr := sdk.MustAccAddressFromBech32(del.DelegatorAddress)
+
+		if err := app.AppKeepers.DistrKeeper.Hooks().BeforeDelegationCreated(ctx, delAddr, valAddr); err != nil {
+			// never called as BeforeDelegationCreated always returns nil
+			panic(fmt.Errorf("error while incrementing period: %w", err))
+		}
+
+		if err := app.AppKeepers.DistrKeeper.Hooks().AfterDelegationModified(ctx, delAddr, valAddr); err != nil {
+			// never called as AfterDelegationModified always returns nil
+			panic(fmt.Errorf("error while creating a new delegation period record: %w", err))
 		}
 	}
 
@@ -152,7 +162,7 @@ func (app *App) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddrs []str
 	counter := int16(0)
 
 	for ; iter.Valid(); iter.Next() {
-		addr := sdk.ValAddress(iter.Key()[1:])
+		addr := sdk.ValAddress(stakingtypes.AddressFromValidatorsKey(iter.Key()))
 		validator, found := app.AppKeepers.StakingKeeper.GetValidator(ctx, addr)
 		if !found {
 			panic("expected validator, not found")
@@ -167,13 +177,14 @@ func (app *App) prepForZeroHeightGenesis(ctx sdk.Context, jailAllowedAddrs []str
 		counter++
 	}
 
-	err := iter.Close()
-	if err != nil {
-		panic(err)
+	if err := iter.Close(); err != nil {
+		app.Logger().Error("error while closing the key-value store reverse prefix iterator: ", err)
+		return
 	}
 
-	if _, err := app.AppKeepers.StakingKeeper.ApplyAndReturnValidatorSetUpdates(ctx); err != nil {
-		panic(err)
+	_, err := app.AppKeepers.StakingKeeper.ApplyAndReturnValidatorSetUpdates(ctx)
+	if err != nil {
+		log.Fatal(err)
 	}
 
 	/* Handle slashing state. */

